### PR TITLE
feat(form): add calendar input field

### PR DESCRIPTION
**Pull request by prudho**
This PR add calendar as a form field, plus it implement its validation and allow to grab underlying value while validating in three different formats.

To do so, a new parameter has been added to the settings: `dateHandling`. It can contain one of these three string:
- `date` (default): field value will be the selected date as a string.
- `input`: field value will be the underlying input value of the calendar.
- `formatter` will let you customize the format of the selected date (or datetime, time, month and year).

Default formatter settings is:
```js
formatter: {
    date: function(date) {
      var date = new Date(date);

      return Intl.DateTimeFormat('en-GB', {
        year: "numeric",
        month: "2-digit",
        day: "2-digit"
      }).format(date);
    },
    datetime: function(date) {
      return Intl.DateTimeFormat('en-GB', {
        year: "numeric",
        month: "2-digit",
        day: "2-digit",
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit'
      }).format(date);
    },
    time: function(date) {
      return Intl.DateTimeFormat('en-GB', {
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit'
      }).format(date);
    },
    month: function(date) {
      return Intl.DateTimeFormat('en-GB', {
        month: '2-digit',
        year: 'numeric'
      }).format(date);
    },
    year: function(date) {
      return Intl.DateTimeFormat('en-GB', {
        year: 'numeric'
      }).format(date);
    }
  }
```

As you can see the default format is the english format, since its [the most used in the world](https://en.wikipedia.org/wiki/Date_format_by_country), but exposing this formatter allow editing it.

Also, the calendar module got a new behavior that return the input text, `get input date`.

### CODE REVIEW

File: src/definitions/behaviors/form.js [hunk 1]
```diff
@@ -660,13 +660,16 @@ $.fn.form = function(parameters) {
...
-                isChecked  = (isCheckbox)
+                $field       = $(field),
+                $parent      = $field.parent(),
+                $grandParent = $parent.parent(),
+                name         = $field.prop('name'),
+                value        = $field.val(),
+                isCheckbox   = $field.is(selector.checkbox),
+                isRadio      = $field.is(selector.radio),
+                isMultiple   = (name.indexOf('[]') !== -1),
+                isCalendar   = ($parent.is(selector.uiCalendar) || $grandParent.is(selector.uiCalendar)),
```
Comment (lubber-de) [inline src/definitions/behaviors/form.js]: What about using the `.closest()` functionality instead? You could get rid of the additional $parent/$grandparent variables and, in case the DOM structure changes, the code is already prepared prepared:

```javascript
$calendar = $field.closest(selector.uiCalendar),
isCalendar   = ($calendar.length>0),
```

File: src/definitions/behaviors/form.js [hunk 2]
```diff
@@ -705,7 +708,60 @@ $.fn.form = function(parameters) {
                       values[name] = false;
                     }
                   }
-                  else {
+                  else if(isCalendar) {
+                    // We must determine which item is the calendar module
+                    var $calendar = $parent.is(selector.uiCalendar)
```
Comment (lubber-de) [inline src/definitions/behaviors/form.js]: If using the above mentioned suggestion, this code can be reduced to `$calendar = $calendar[0] || false;`

File: src/definitions/behaviors/form.js [hunk 3]
```diff
@@ -1486,6 +1544,46 @@ $.fn.form.settings = {
     }
   },
 
+  formatter: {
+    date: function(date) {
+      var date = new Date(date);
+
+      return Intl.DateTimeFormat('en-GB', {
+        year: "numeric",
```
Comment (lubber-de) [inline src/definitions/behaviors/form.js]: Not needed. en-GB returns d/m/y by default

```javscript
return Intl.DateTimeFormat('en-GB').format(date);
```
is sufficient

**Comment by lubber-de** [comment]
I think the `reset` and `clear` functions have to be adjusted as well, because they will empty the form field, but the underlying date within the calendar module stays untouched

**Comment by lubber-de** [comment]
What about adding rules for the date field like `time`, `date`, `datetime` ? You added the format to the `get values` behavior / function, but the only possible validation would be `empty`

Or did i miss something? :thinking:

**Comment by lubber-de** [comment]
`Intl.DateTimeFormat.protoype.format()` is not available for Android Webview (but `formatToParts` is).
Here is a simple polyfill, which we might add.
```javascript
    if(!('format' in Intl.DateTimeFormat.prototype)){
        Intl.DateTimeFormat.prototype.format = function(date){
            return this.formatToParts(date).map(function(o){return o.value}).join('');
        }
    }
```

File: src/definitions/behaviors/form.js (hunk 1 above)
Comment (prudho) [reply src/definitions/behaviors/form.js]: Wow good catch here ! I think that I have over complicated the code 👍

File: src/definitions/behaviors/form.js (hunk 2 above)
Comment (prudho) [reply src/definitions/behaviors/form.js]: With the simplification, I think that this piece of code is now useless...

File: src/definitions/behaviors/form.js (hunk 3 above)
Comment (prudho) [reply src/definitions/behaviors/form.js]: True. I really need some sleep !

**Comment by prudho** [comment]
> I think the `reset` and `clear` functions have to be adjusted as well, because they will empty the form field, but the underlying date within the calendar module stays untouched

Yup, this is now done.

> What about adding rules for the date field like `time`, `date`, `datetime` ? You added the format to the `get values` behavior / function, but the only possible validation would be `empty`
> 
> Or did i miss something? 🤔

Yes, actually providing a validation other than `empty` is pretty hard to put up. The returned value depends on the type of calendar and the selected `dateHandling` parameter and defining rule for each case would be a real PITA. Plus don't forget that the calendar module already return a formatted value, so why re-validating it ?

**Comment by lubber-de** [comment]
> Plus don't forget that the calendar module already return a formatted value, so why re-validating it ?
OK, then let's leave it for now 🙂

File: src/definitions/behaviors/form.js [hunk 4]
```diff
@@ -1486,6 +1540,42 @@ $.fn.form.settings = {
     }
   },
 
+  formatter: {
+    date: function(date) {
+      var date = new Date(date);
```
Comment (lubber-de) [inline src/definitions/behaviors/form.js]: Why is this necessary here? It's not used in the other formatters? :thinking:

Comment (prudho) [reply src/definitions/behaviors/form.js]: Mmmmh. I defined it during my tests, and did not implement it in other formatters. The `Intl.DateTimeFormat.prototype.format()` seem to need a date as input, but looks like he perfectly handle a date string. So, should I remove `var date = new Date(date);` from this one or implement a Date object in other ?

Comment (lubber-de) [reply src/definitions/behaviors/form.js]: In Line 721 you added 
```javascript
var date = $calendar.calendar('get date');
```
which will already return a valid Date Object (or null/undefined) and is passed to all the formatters, so you can remove the `var date=new Date(date)`

Comment (lubber-de) [reply src/definitions/behaviors/form.js]: As said, you can safely remove the ´var date=new Date(date)` line.
I would love to approve this PR right after that 😉

Comment (prudho) [reply src/definitions/behaviors/form.js]: Done.

**Comment by lubber-de** [comment]
Docs added by https://github.com/fomantic/Fomantic-UI-Docs/pull/131
